name: Advanced AI model training

on:
  push:
    branches: [main]
  pull_request:
    branches: [main]
  schedule:
    - cron: '0 17 * * *'

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3
      - name: Set up Python 3.11
        uses: actions/setup-python@v3
        with:
          python-version: 3.11
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install -r requirements.txt
      - name: Download data
        run: |
          python src/prepare_data.py
      - name: Validate data
        run: |
          python src/validate_data.py
      - name: Train model
        run: |
          python src/model.py
      - name: Validate model
        run: |
          python src/validate_model.py
      - name: Run tests
        run: |
          python -m unittest tests/test_model.py
      - name: Archive model
        uses: actions/upload-artifact@v2
        with:
          name: model
          path: model.pkl
      - name: Notify
        if: failure()
        uses: octokit/request-action@v2.x
        with:
          route: POST /repos/{owner}/{repo}/issues
          title: 'Model training failed ⚠️'
          body: 'The model training action failed. Please see the logs for more information. ⚠️'
          labels: 'bug'
          owner: PunGrumpy
          repo: machine-learning-python
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
